drv/stm32xx-gpio-common/src/server.rs

```diff
@@ -0,0 +1,278 @@
+    fn toggle(&self, pins: u16);
+    fn read(&self) -> u16;
+}
+
+/// Adapter from `GpioPeriph`, the trait implemented for disjoint GPIO
+/// peripheral types, to the general `AnyGpioPeriph`.
+#[cfg(feature = "server-support")]
+impl<T: GpioPeriph> AnyGpioPeriph for T {
+    fn configure(&self, pins: u16, atts: u16) {
+        // The GPIO config registers come in 1, 2, and 4-bit per field variants.
+        // The user-submitted mask is already correct for the 1-bit fields; we
+        // need to expand it into corresponding 2- and 4-bit masks. We use an
+        // outer perfect shuffle operation for this, which interleaves zeroes
+        // from the top 16 bits into the bottom 16.
+
+        // 1 in each targeted 1bit field.
+        let mask_1 = u32::from(pins);
+        // 0b01 in each targeted 2bit field.
+        let lsbs_2 = outer_perfect_shuffle(mask_1);
+        // 0b0001 in each targeted 4bit field for low half.
+        let lsbs_4l = outer_perfect_shuffle(lsbs_2 & 0xFFFF);
+        // Same for high half.
+        let lsbs_4h = outer_perfect_shuffle(lsbs_2 >> 16);
+
+        // Corresponding masks, with 1s in all field bits instead of just the
+        // LSB.  We use multiplication to distribute a constant over all 1
+        // positions because multiplication is cheap on ARMvx-M devices.
+        let mask_2 = lsbs_2.wrapping_mul(0b11);
+        let mask_4l = lsbs_4l.wrapping_mul(0b1111);
+        let mask_4h = lsbs_4h.wrapping_mul(0b1111);
+
+        // MODER contains 16x 2-bit fields.
+        let moder_val = u32::from(atts & 0b11);
+        self.moder().modify(|r, w| unsafe {
+            w.bits((r.bits() & !mask_2) | (moder_val.wrapping_mul(lsbs_2)))
+        });
+
+        // OTYPER contains 16x 1-bit fields.
+        let otyper_val = u32::from((atts >> 2) & 1);
+        self.otyper().modify(|r, w| unsafe {
+            w.bits((r.bits() & !mask_1) | (otyper_val.wrapping_mul(mask_1)))
+        });
+        // OSPEEDR contains 16x 2-bit fields.
+        let ospeedr_val = u32::from((atts >> 3) & 0b11);
+        self.ospeedr().modify(|r, w| unsafe {
+            w.bits((r.bits() & !mask_2) | (ospeedr_val.wrapping_mul(lsbs_2)))
+        });
+        // PUPDR contains 16x 2-bit fields.
+        let pupdr_val = u32::from((atts >> 5) & 0b11);
+        self.pupdr().modify(|r, w| unsafe {
+            w.bits((r.bits() & !mask_2) | (pupdr_val.wrapping_mul(lsbs_2)))
+        });
+        // AFRx contains 8x 4-bit fields.
+        let af_val = u32::from((atts >> 7) & 0b1111);
+        self.afrl().modify(|r, w| unsafe {
+            w.bits((r.bits() & !mask_4l) | (af_val.wrapping_mul(lsbs_4l)))
+        });
+        self.afrh().modify(|r, w| unsafe {
+            w.bits((r.bits() & !mask_4h) | (af_val.wrapping_mul(lsbs_4h)))
+        });
+    }
+
+    fn set_reset(&self, set: u16, reset: u16) {
+        self.bsrr().write(|w| unsafe {
+            w.bits((u32::from(reset) << 16) | u32::from(set))
```

Comment:
General question: why use `u32::from` vs `.into()`?

Reply:
Because it's explicit about the destination type, which makes the code easier for me to read. (I habitually use `T::from` over `.into()` for this reason. This is possibly idiosyncratic of me.)

Reply:
Makes sense (and there was follow-up discussion in #oxide-rust)